xapian-core/common/serialise-double.cc

```diff
@@ -38,175 +40,36 @@
-
-#if FLT_RADIX == 2
-# define MAX_MANTISSA_BYTES ((DBL_MANT_DIG + 7 + 7) / 8)
-# define MAX_EXP ((DBL_MAX_EXP + 1) / 8)
-# define MAX_MANTISSA (1 << (DBL_MAX_EXP & 7))
-#elif FLT_RADIX == 16
-# define MAX_MANTISSA_BYTES ((DBL_MANT_DIG + 1 + 1) / 2)
-# define MAX_EXP ((DBL_MAX_EXP + 1) / 2)
-# define MAX_MANTISSA (1 << ((DBL_MAX_EXP & 1) * 4))
-#else
-# error FLT_RADIX is a value not currently handled (not 2 or 16)
-// # define MAX_MANTISSA_BYTES (sizeof(double) + 1)
-#endif
-
-static int base256ify_double(double &v) {
-    int exp;
-    v = frexp(v, &exp);
-    // v is now in the range [0.5, 1.0)
-    --exp;
-#if FLT_RADIX == 2
-    v = scalbn(v, (exp & 7) + 1);
-#else
-    v = ldexp(v, (exp & 7) + 1);
+string serialise_double(double v)
+{
+    size_t double_size = sizeof(double);
+#ifdef WORDS_BIGENDIAN
+    uint64_t temp;
+    static_assert(sizeof(temp) == sizeof(v));
+    memcpy(&temp, &v, double_size);
+    temp = do_bswap(temp);
+    return string(reinterpret_cast<const char *>(&temp), double_size);
 #endif
-    // v is now in the range [1.0, 256.0)
-    exp >>= 3;
-    return exp;
+    return string(reinterpret_cast<const char *>(&v), double_size);
 }
 
-std::string serialise_double(double v)
+double unserialise_double(const char ** p, const char * end)
 {
-    /* First byte:
-     *  bit 7 Negative flag
-     *  bit 4..6 Mantissa length - 1
-     *  bit 0..3 --- 0-13 -> Exponent + 7
-     *            \- 14 -> Exponent given by next byte
-     *             - 15 -> Exponent given by next 2 bytes
-     *
-     * Then optional medium (1 byte) or large exponent (2 bytes, lsb first)
-     *
-     * Then mantissa (0 iff value is 0)
-     */
-
-    bool negative = (v < 0.0);
-
-    if (negative) v = -v;
-
-    int exp = base256ify_double(v);
-
-    string result;
-
-    if (exp <= 6 && exp >= -7) {
-	unsigned char b = static_cast<unsigned char>(exp + 7);
-	if (negative) b |= static_cast<unsigned char>(0x80);
-	result += char(b);
-    } else {
-	if (exp >= -128 && exp < 127) {
-	    result += negative ? char(0x8e) : char(0x0e);
-	    result += char(exp + 128);
-	} else {
-	    if (exp < -32768 || exp > 32767) {
-		throw Xapian::InternalError("Insane exponent in floating point number");
-	    }
-	    result += negative ? char(0x8f) : char(0x0f);
-	    result += char(unsigned(exp + 32768) & 0xff);
-	    result += char(unsigned(exp + 32768) >> 8);
-	}
-    }
-
-    int maxbytes = min(MAX_MANTISSA_BYTES, 8);
-
-    size_t n = result.size();
-    do {
-	unsigned char byte = static_cast<unsigned char>(v);
-	result += char(byte);
-	v -= double(byte);
-	v *= 256.0;
-    } while (v != 0.0 && --maxbytes);
-
-    n = result.size() - n;
-    if (n > 1) {
-	Assert(n <= 8);
-	result[0] = static_cast<unsigned char>(result[0] | ((n - 1) << 4));
+    if (end - *p < 8) {
+		throw Xapian::SerialisationError(
+				"Bad encoded double: insufficient data");
     }
-
+    size_t double_size = sizeof(double);
+    double result;
+    memcpy(&result, *p, double_size);
```

Comment:
The above line is OK (`double_size` concerns aside) for the little-endian case, but in the big-endian case you copy the data from p into a `double` here, then below copy it from there into a `uint64_t`, then byte swap that and, then copy back to the `double`.

That first copy is an extra copy you don't need.  Maybe the compiler will be able to optimise it away, but if you eliminate it then you don't need to worry about whether it will.

And that first copy to a `double` seems icky as the value being copied is not really a bit pattern for a `double` as such - we know reversing the bytes gives a `double` value.  In practice that probably won't cause problems, but it's conceivable it could be some sort of trapping value and trigger a floating point exception if the CPU loads it.  Again if we avoid this copy, that's something we can avoid having to worry about.

Reply:
Ah yes, hadn't thought about the second possibility. Will fix it right away.